sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusReceiverClient.java

```diff
@@ -624,23 +623,26 @@ public void close() {
      * entity.
      */
     private void queueWork(int maximumMessageCount, Duration maxWaitTime,
-        FluxSink<ServiceBusReceivedMessageContext> emitter) {
-        synchronized (lock) {
+            FluxSink<ServiceBusReceivedMessageContext> emitter) {
             final long id = idGenerator.getAndIncrement();
-            EmitterProcessor<ServiceBusReceivedMessageContext> emitterProcessor = messageProcessor.get();
-
             final SynchronousReceiveWork work = new SynchronousReceiveWork(id, maximumMessageCount, maxWaitTime,
                 emitter);
-            final SynchronousMessageSubscriber syncSubscriber = new SynchronousMessageSubscriber(work);
-            logger.info("[{}]: Started synchronous message subscriber.", id);
 
-            if (emitterProcessor == null) {
-                emitterProcessor = this.asyncClient.receive(DEFAULT_RECEIVE_OPTIONS)
-                    .subscribeWith(EmitterProcessor.create(asyncClient.getReceiverOptions().getPrefetchCount(), false));
-                messageProcessor.set(emitterProcessor);
+            SynchronousMessageSubscriber messageSubscriber = synchronousMessageSubscriber.get();
+            if (messageSubscriber == null) {
+                long prefetch = asyncClient.getReceiverOptions().getPrefetchCount();
+                SynchronousMessageSubscriber newSubscriber = new SynchronousMessageSubscriber(prefetch, work);
+
+                if (!synchronousMessageSubscriber.compareAndSet(null, newSubscriber)) {
+                    newSubscriber.dispose();
+                    SynchronousMessageSubscriber existing = synchronousMessageSubscriber.get();
+                    existing.queueWork(work);
+                } else {
+                    asyncClient.receive(DEFAULT_RECEIVE_OPTIONS).subscribeWith(newSubscriber);
```

Comment:
When a `newSubscriber` instance is created , we queue the work.. We can not do after `subscribeWith` is called because you may receive a message before you have anything in the queue.